lib/edamam.rb

```diff
@@ -0,0 +1,192 @@
+        &.dig('hints')
+        &.first
+
+      return unless upc_lookup_data&.is_a?(Hash)
+
+      food_data = upc_lookup_data['food']
+      return unless food_data&.is_a?(Hash)
+
+      food_id = food_data['foodId']
+      name = food_data['label']
+      return unless food_id && name
+
+      measurement_data = upc_lookup_data['measures']
+      measurement_units =
+        if measurement_data&.is_a?(Array)
+          measurement_data.map { |measure| measure['label'] if measure&.is_a?(Hash) }.compact
+        else
+          []
+        end
+
+      body = {
+        ingredients: [
+          {
+            quantity: 1,
+            measureURI: "#{MEASUREMENT_URI_PREFIX}tablespoon",
+            foodId: food_id
+          }
+        ]
+      }
+      request_data = {
+        body: body.to_json,
+        headers: { 'Content-Type': 'application/json' },
+        method: :post,
+        name: 'nutrition data',
+        url: NUTRITION_DATA_URL
+      }
+      response = send_request(request_data)
+
+      grams_per_tablespoon = get_weight_from_nutrition_data_response(response)
+      unless grams_per_tablespoon
+        body[:ingredients].first[:measureURI] = "#{MEASUREMENT_URI_PREFIX}gram"
+        request_data[:body] = body.to_json
+        response = send_request(request_data)
+      end
+
+      nutrients = response['totalNutrients']
+      return unless nutrients
+
+      divisor = grams_per_tablespoon || 1
+      nutrients_data = nutrient_mappings.map do |nutrient, code|
+        nutrient_data = nutrients[code]
+        value =
+          if nutrient_data
+            case nutrient_data['unit']
+            when 'g'
+              cast_decimal(nutrient_data['quantity']) / divisor
+            when 'mg'
+              cast_decimal(nutrient_data['quantity']) / divisor * 1_000
+            when 'µg'
+              cast_decimal(nutrient_data['quantity']) / divisor * 1_000_000
+            else
+              nil
+            end
+          end
+
+        [nutrient, value]
+      end.to_h.merge(calories: cast_decimal(nutrients.dig('ENERC_KCAL', 'quantity')))
+
+      {
+        food_id: food_id,
+        grams_per_tablespoon: grams_per_tablespoon,
+        health_labels: response['healthLabels'],
+        measurement_units: measurement_units,
+        name: name,
+        nutrients: nutrients_data
+      }
+    end
+
+    private
+
+    def cast_decimal(value)
+      Kernel.Float(value).to_d
+    rescue ArgumentError, TypeError
+      nil
+    end
+
+    def get_weight_from_nutrition_data_response(data)
+      return unless data&.is_a?(Hash)
+
+      data = data['ingredients']
+      return unless data&.is_a?(Array)
+
+      data = data.first
+      return unless data&.is_a?(Hash)
+
+      data = data['parsed']
+      return unless data&.is_a?(Array)
+
+      data = data.first
+      return unless data&.is_a?(Hash)
+
+      data['weight'] unless data['status'] == 'MISSING_QUANTITY'
```

Comment:
🤯 How could Mr. I-love-dig write such a method⁉️

Reply:
Dig doesn't check anything other than values being `nil`. And it isn't just that an error could be raised if the response isn't formatted how we expect because hashes, arrays, strings, and integers all have a `[]` method, and all of them except for integers also have a `first` method. So if the response wasn't formatted correctly this method could potentially do some weird stuff otherwise.

Reply:
Well, none of that matters when the last property we dig is `weight` and that would never exist if the data wasn't valid. But also, we shouldn't be guarding random broken data coming back.

Reply:
maybe I just don't trust third parties enough 😛